jobsdb/jobsdb.go

```diff
@@ -2031,9 +2043,24 @@ func (jd *HandleT) WithStoreSafeTx(f func(tx StoreSafeTx) error) error {
 
 func (jd *HandleT) inStoreSafeCtx(f func() error) error {
 	// Only locks the list
-	jd.dsListLock.RLock()
-	defer jd.dsListLock.RUnlock()
-	return f()
+	op := func() error {
+		jd.dsListLock.RLock()
+		defer jd.dsListLock.RUnlock()
+		return f()
+	}
+	var err error
+	for {
+		err = op()
+		if err != nil && errors.Is(err, errStaleDsList) {
+			jd.logger.Errorf("[JobsDB] :: Store failed: %v. Retrying after refreshing DS cache", errStaleDsList)
+			jd.dsListLock.WithLock(func(l lock.DSListLockToken) {
+				_ = jd.refreshDSList(l)
+			})
+		} else {
+			break
+		}
+	}
```

Comment:
In case of high load or when `maxDSSize` is less, there is a possibility that even the newly added DS gets filled by the time we refresh & try again. So, we keep refreshing DS until `Store()` succeeds or some other error is returned. 
And, given we call it with `Context.WithTimeout()` it will return in bounded time.